### PR TITLE
Task Module

### PR DESCRIPTION
**Pull request by MarkHaoxiang**
Files to focus on:
- nmmo/task/task_api.py
- nmmo/task/predicate/core.py
- nmmo/task/game_state.py
Examples of predicate definition can be found in other files inside predicate folder.

Completed:
- Simple function definition for predicates
- Predicate API expects a float return value
- Refactored Group and moved some files around.
- Refactored TaskEnv to expect and Task to follow some of the points discussed last week and in the document. 

TODO before merge:
- GroupObsView: A wrapper on Observations for ease of access
- Finish refactoring the remaining predicates
- Examples of non-predicate based task (?)
- TaskSampler definition (?)
- Rewrite predicates to return completion instead of binary 1.0/0.0
- Rewrite predicate tests to consider float returns. 
- Add more tests for the new environment wrapper.
- Test for correct caching optimization behavior.

### CODE REVIEW

File: nmmo/task/predicate/core.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
+from abc import ABC, abstractmethod
+from typing import Dict, List, Union
+
+from nmmo.task.game_state import GameState
+from nmmo.task.group import Group
+
+class Predicate(ABC):
+  ''' A mapping from the state of an episode to a boolean
```
Comment (RyanNavillus) [inline nmmo/task/predicate/core.py]:
```suggestion
  ''' A mapping from the state of an episode to a boolean or float
```

File: nmmo/task/task_api.py [hunk 2]
```diff
@@ -1,81 +1,146 @@
...
+  def rewards(self, gs: GameState):
+    rewards = {int(ent_id): 0 for ent_id in self._assignee}
+    infos = {int(ent_id): {self._predicate.name: self.evaluate(gs)}
+             for ent_id in self._assignee}
+    if self.evaluate(gs):
+      rewards = {int(ent_id): self._reward for ent_id in self._assignee}
+    return rewards, infos
 
+class MultiTask(Task):
+  def __init__(self, *tasks: Task):
```
Comment (RyanNavillus) [inline nmmo/task/task_api.py]: Might make sense to allow you to weight tasks, so provide a list of task weights or make tasks a dictionary of {task:weight}

File: nmmo/task/task_api.py [hunk 3]
```diff
@@ -1,81 +1,146 @@
...
+################################################
+# Environment Wrapper
+# Eventually should replace env.py once stable
+# TODO(mark) Syllabus
 
-# TODO(kywch): how to divide the task system, into core vs. wrapper, etc.?
-#  We will address this once the task wrapper is stable
 # pylint: disable=abstract-method
-class TaskWrapper(nmmo.Env):
+class TaskEnv(nmmo.Env):
```
Comment (RyanNavillus) [inline nmmo/task/task_api.py:108]: If you want to use Syllabus you can make this subclass TaskWrapper instead. Alternatively, maybe I can just make a TaskEnv mixin for Syllabus That basically does what you have here.

**Comment by RyanNavillus** [comment]
This looks pretty good. I think that having predicates be classes, but allowing people to define them with simple functions provides the most simplicity and flexibility. Your generic Task and specific Predicate structure looks very similar to what I came up with so I like the direction there. MultiTask looks like it provides enough functionality to do a lot of cool tasks.
Questions:
* The code seems inconsistent about whether predicate outputs will be boolean or float. Are we agreed on using float values for this?
* Can predicates call other predicates? Do you have an example of a composite predicate like that?
    * I think it'll be valuable for ELM if we compose the predicates whenever possible. Having simple atomic predicates to work with should make it easier to generate new ones. It might also make caching more effective.
* What does ItemPredicate do? Do you need to define another @predicate(item) or @item_predicate for that?

File: nmmo/task/task_api.py (hunk 2 above)
Comment (MarkHaoxiang) [reply nmmo/task/task_api.py]: Good idea, added an option to pass a task into Multitask as (task, weight).

File: nmmo/task/predicate/core.py (hunk 1 above)
Comment (MarkHaoxiang) [reply nmmo/task/predicate/core.py]: Clarified as a float in range [0,1] where True maps to 1.0 and False 0.0.

**Comment by MarkHaoxiang** [comment]
1) Have clarified in core that predicates map to float, where True is 1.0 and False is 0.0. Will start translating some individual predicates to return a completion percentage where possible.

2) Yes, that was the original intention. Will refactor `@predicate` to better support this type of behavior and provide some examples.

3)  Depreciated and refactored out.

**Comment by MarkHaoxiang** [comment]
**Implementation of the task system.** 

Primary changes be found in the task folder https://github.com/CarperAI/nmmo-environment/tree/task-0.2/nmmo/task, as well as tests and examples https://github.com/CarperAI/nmmo-environment/tree/task-0.2/tests/task. 

Primary changes are
1. Definition of predicates, as well as corresponding API for returning a floating point value.
2. Redefined Group
3. Syntactic sugar in Game State
4. Callable task class with baselines using predicates
5. TaskEnv
6. Related tests

The PR is quite large as it also contains tooling and cleanup from the branch event-log.

**Comment by MarkHaoxiang** [comment]
Closing pr Changes will be rebased on v1.6 and resubmitted. Also need to work on putting task embedding info into observations.

**Pull request closed without merging**